apiserver/apiserver/coordinator/matchmaking.py

```diff
@@ -129,9 +135,114 @@ def serve_game_task(conn, has_gpu=False):
             "width": map_width,
             "height": map_height,
             "users": players,
+            "challenge": None,
         })
 
 
+def reset_challenges(conn):
+    """Check ongoing challenges, and reset ones that are "stuck"."""
+    reset_stuck_challenges = model.challenges.update().where(
+        (model.challenges.c.status == model.ChallengeStatus.PLAYING_GAME.value) &
+        (model.challenges.c.most_recent_game_task <
+         datetime.datetime.now() - datetime.timedelta(
+             minutes=30))
+    ).values(
+        status=model.ChallengeStatus.CREATED.value,
+    )
+    conn.execute(reset_stuck_challenges)
+
+
+def find_challenge(conn, has_gpu=False):
+    """
+    Find a set of players in a challenge.
+    """
+    # Reset any stuck challenges
+    reset_challenges(conn)
+
+    challenge = conn.execute(
+        model.challenges.select(
+            model.challenges.c.status == model.ChallengeStatus.CREATED.value
+        ).order_by(
+            model.challenges.c.most_recent_game_task.asc()
+        )
+    ).first()
+
+    if not challenge:
+        return None
+
+    player_filter = sqlalchemy.sql.expression.true()
+    if not has_gpu:
+        player_filter = (model.ranked_bots_users.c.is_gpu_enabled == False)
+
+    total_players = conn.execute(model.total_ranked_users).first()[0]
+    thresholds = util.tier_thresholds(total_players)
+    ranked_users = model.ranked_users_query()
+    bots_query = sqlalchemy.sql.select([
+        model.ranked_bots_users.c.user_id,
+        model.ranked_bots_users.c.bot_id,
+        model.ranked_bots_users.c.username,
+        model.ranked_bots_users.c.rank,
+        model.ranked_bots_users.c.num_submissions.label("version_number"),
+    ]).select_from(
+        model.ranked_bots_users.join(
+            model.bots,
+            (model.ranked_bots_users.c.user_id == model.bots.c.user_id) &
+            (model.ranked_bots_users.c.bot_id == model.bots.c.id)
+        )
+    ).where(
+        (model.bots.c.compile_status == model.CompileStatus.SUCCESSFUL.value) &
+        sqlalchemy.sql.exists(
+            model.challenge_participants.select(
+                (model.ranked_bots.c.user_id == model.challenge_participants.c.user_id) &
+                (model.challenge_participants.c.challenge_id == challenge["id"]) &
+                (model.challenge_participants.c.user_id == model.ranked_bots_users.c.user_id)
+            )
+        ) &
+        player_filter
+    )
+    bots = conn.execute(bots_query).fetchall()
+
+    # TODO: assumes one-bot-per-player
+    if len(bots) < 2:
+        return None
+
+    selected_bots = []
+    candidate_bots = []
+    for bot in bots:
+        if bot["user_id"] == challenge["issuer"]:
+            selected_bots.append(bot)
+        else:
+            candidate_bots.append(bot)
+
+    if random.random() < 0.5 and len(candidate_bots) == 3:
```

Comment:
It seems like this is assuming the number of challenge participants will either be 2 or 4. This can be easily modified to handle an arbitrary number though.

    if random.random() < 0.5 and len(candidate_bots) >= 3:
        selected_bots.extend(random.shuffle(candidate_bots)[:3])